compiler/rustc_ast_lowering/src/item.rs

```diff
@@ -1379,58 +1379,66 @@ impl<'hir> LoweringContext<'_, 'hir> {
-                                 point where a type parameter is declared",
-                            );
-                        };
-                        // Check if the where clause type is a plain type parameter.
-                        match bound_pred.bounded_ty.kind {
-                            TyKind::Path(None, ref path)
-                                if path.segments.len() == 1
-                                    && bound_pred.bound_generic_params.is_empty() =>
-                            {
-                                if let Some(Res::Def(DefKind::TyParam, def_id)) = self
-                                    .resolver
-                                    .get_partial_res(bound_pred.bounded_ty.id)
-                                    .map(|d| d.base_res())
-                                {
-                                    if let Some(def_id) = def_id.as_local() {
-                                        for param in &generics.params {
-                                            if let GenericParamKind::Type { .. } = param.kind {
-                                                if def_id == self.resolver.local_def_id(param.id) {
-                                                    add_bounds
-                                                        .entry(param.id)
-                                                        .or_default()
-                                                        .push(bound.clone());
-                                                    continue 'next_bound;
-                                                }
-                                            }
-                                        }
-                                    }
-                                }
-                                report_error(self)
-                            }
-                            _ => report_error(self),
+            let bound_pred = match *pred {
+                WherePredicate::BoundPredicate(ref bound_pred) => bound_pred,
+                _ => continue,
+            };
+            'next_bound: for bound in &bound_pred.bounds {
+                if !matches!(*bound, GenericBound::Trait(_, TraitBoundModifier::Maybe)) {
+                    continue;
+                }
+                // Check if the where clause type is a plain type parameter.
+                if let Some(def_id) = self.get_def_id(bound_pred) {
+                    // Search for it in the generic type parameters.
+                    for param in &generics.params {
+                        if !matches!(param.kind, GenericParamKind::Type { .. }) {
+                            continue;
+                        }
+                        if def_id == self.resolver.local_def_id(param.id) {
+                            add_bounds.entry(param.id).or_default().push(bound.clone());
+                            continue 'next_bound;
                         }
                     }
```

Comment:
This loop can be refactored into a call to `generics.params.iter().find`. The `matches!` test should perhaps be an assertion on the found parameter.

Reply:
Are you saying that `param.id` won't match unless it's a type parameter? I guess duplicate `LocalDefId`s aren't supposed to happen, and we already know that `def_id` is for a `TyParam`. I'm also wary of adding an assertion where there wasn't one in the original code.

On the other hand, the `matches!` will allow skipping of calls to `self.resolver.local_def_id` for non-type generic parameters, which could have a performance impact, given that this function is kind of quadratic in the number of generic parameters.